editoast/src/views/infra/mod.rs

```diff
@@ -141,6 +142,58 @@ async fn create(
     .await
 }
 
+/// Duplicate an infra
+#[post("/<infra>/clone")]
+async fn clone(infra: i64, conn: DBConnection) -> ApiResult<Custom<Json<Infra>>> {
+    conn.run(move |conn| {
+        let cloned_infra = Infra::clone(infra, conn)?;
+        let model_list = [
+            "osrd_infra_tracksectionmodel",
+            "osrd_infra_signalmodel",
+            "osrd_infra_speedsectionmodel",
+            "osrd_infra_tracksectionlinkmodel",
+            "osrd_infra_switchmodel",
+            "osrd_infra_switchtypemodel",
+            "osrd_infra_detectormodel",
+            "osrd_infra_bufferstopmodel",
+            "osrd_infra_routemodel",
+            "osrd_infra_operationalpointmodel",
+            "osrd_infra_catenarymodel",
+        ];
+        let generated_list = [
+            "osrd_infra_tracksectionlayer",
+            "osrd_infra_signallayer",
+            "osrd_infra_speedsectionlayer",
+            "osrd_infra_tracksectionlinklayer",
+            "osrd_infra_switchlayer",
+            "osrd_infra_detectorlayer",
+            "osrd_infra_bufferstoplayer",
+            "osrd_infra_operationalpointlayer",        
+            "osrd_infra_catenarylayer",];
+
+
+        for table in model_list {
+            sql_query(format!(
+                "INSERT INTO {table}(id, obj_id,data,infra_id) SELECT nextval('{table}_id_seq'), obj_id,data,$1 FROM {table} WHERE infra_id=$2"
+            ))
+            .bind::<BigInt, _>(cloned_infra.id)
+            .bind::<BigInt, _>(infra)
+            .execute(conn)?;
+        }
+
+        for table in generated_list {
+            sql_query(format!(
+                "INSERT INTO {table}(id, obj_id,geographic,schematic,infra_id) SELECT nextval('{table}_id_seq'), obj_id,geographic,schematic,$1 FROM {table} WHERE infra_id=$2"
+            ))
+            .bind::<BigInt, _>(cloned_infra.id)
+            .bind::<BigInt, _>(infra)
+            .execute(conn)?;
+        }
+
+        Ok(Custom(Status::Created, Json(cloned_infra)))
+    })
+    .await
```

Comment:
You should do `conn.run` inside the loops then join the future so it's run in parallel.